Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,556 @@
+    public struct Writer: ~Copyable {
+        /// Writes data to the response stream
+        public func write(_ byteBuffer: ByteBuffer) async throws
+        /// Closes off the response stream
+        public func finish() async throws
+        /// Writes the `byteBuffer` to the response stream and subsequently closes the stream
+        public func writeAndFinish(_ byteBuffer: ByteBuffer) async throws
+    }
+
+    /// Creates an empty lambda response
+    public init()
+
+    /// Creates a LambdaResponse with a fixed ByteBuffer.
+    public init(_ byteBuffer: ByteBuffer)
+
+    /// Creates a streamed lambda response. Use the ``Writer`` to send
+    /// response chunks on the stream.
+    public init(_ stream: @escaping sending (Writer) async throws -> ())
+}
+```
+
+The `StreamingLambdaHandler` would look like this:
+
+```swift
+/// The base LambdaHandler protocol
+public protocol StreamingLambdaHandler: ~Copyable {
+    /// The business logic of the Lambda function
+    /// - Parameters:
+    ///   - event: The invocation's input data
+    ///   - context: The LambdaContext containing the invocation's metadata
+    /// - Returns: A LambdaResponse, that can be streamed
+    mutating func handle(
+        _ event: ByteBuffer,
+        context: LambdaContext
+    ) async throws -> LambdaResponse
+}
+```
+
+There are pros and cons for the API that returns the `LambdaResponses` and there are pros and cons for the API that receives a `LambdaResponseWriter` as a parameter.
+
+Concerning following structured concurrency principles the approach that receives a `LambdaResponseWriter` as a parameter has benefits as the lifetime of the handle function is tied to the invocation runtime.
+The approach that returns a `LambdaResponse` splits the invocation into two separate function calls. First the handle method is invoked, second the `LambdaResponse` writer closure is invoked. This means that it is impossible to use Swift APIs that use `with` style lifecycle management patterns from before creating the response until sending the full response stream off. This means for example: That users instrumenting their lambdas with Swift tracing likely can not use the `withSpan` API for the full lifetime of the request, if they return a streamed response.
+
+However, if it comes to consistency with the larger Swift on server ecosystem, the API that returns a `LambdaResponse` is likely the better choice. Hummingbird v2, OpenAPI and the new Swift gRPC v2 implementation all use this approach. This might be due to the fact that writing middleware becomes easier, if a Response is explicitly returned.
+
+We decided to implement the approach in which a `LambdaResponseWriter` is passed to the function, since the approach in which a `LambdaResponse` is returned can trivially be built on top of it. This is not true vice versa.
+
+We welcome the discussion on this topic and are open to change our minds and API here.
```

Comment:
It would be interesting to see if anyone has any use cases that integrate something like Swift Middleware

Reply:
I think this is a fair point and I would love to see how middleware would work with this writer approach. IMO middleware do make some sense for Lambda

Reply:
I made this point above, middleware definitely does make sense for lambda (for example tracing should just be a middleware that can be used anywhere) and should just be integrated IMO. My only quasi-concern with the passed-in-writer-approach is that for it to work with middleware, the middleware protocol has to suppress `Copyable` conformance on one of its associated types - its writer type, something that isn't going to be supported in Swift 6.0.

That said I do think the argument that this is a lower-level and more flexible API makes sense.

Reply:
I don't think we want to integrate the swift-middleware library directly. I think if users have those needs they should consider using hummingbird/smoke/vapor in lambda mode, which then brings the option to use middleware.

This shall not mean that writing a middleware can be impossible. We should allow users to write middleware that they can stack. 

Also I think swift-middleware is currently not targeted for anytime soon. So in order to make progress we should not depend on it. Integrating swift-middleware (as is) on top of this proposed API is absolutely possible and therefore should not be considered a blocker.

Reply:
We can debate whether we think the swift-middleware library should be integrated directly but I do agree it shouldn't be considered a blocker for this work/release.